bosh-monitor/lib/bosh/monitor/instance_manager.rb

```diff
@@ -0,0 +1,359 @@
+        plugin_class = Bosh::Monitor::Plugins.const_get(class_name)
+      rescue NameError => e
+        raise PluginError, "Cannot find '#{name}' plugin"
+      end
+
+      plugin_class.new(options)
+    end
+
+    def setup_events
+      @processor.enable_pruning(Bhm.intervals.prune_events)
+      Bhm.plugins.each do |plugin|
+        @processor.add_plugin(lookup_plugin(plugin["name"], plugin["options"]), plugin["events"])
+      end
+
+      EM.schedule do
+        Bhm.nats.subscribe("hm.agent.heartbeat.*") do |message, reply, subject|
+          process_event(:heartbeat, subject, message)
+        end
+
+        Bhm.nats.subscribe("hm.agent.alert.*") do |message, reply, subject|
+          process_event(:alert, subject, message)
+        end
+
+        Bhm.nats.subscribe("hm.agent.shutdown.*") do |message, reply, subject|
+          process_event(:shutdown, subject, message)
+        end
+      end
+    end
+
+    def agents_count
+      agents = Set.new(@rogue_agents.keys)
+      agents.merge(all_managed_agent_ids)
+      agents.size
+    end
+
+    def deployments_count
+      @deployment_name_to_deployments.size
+    end
+
+    # Syncs deployments list received from director
+    # with HM deployments.
+    # @param deployments Array list of deployments returned by director
+    def sync_deployments(deployments)
+      active_deployment_names = sync_active_deployments(deployments)
+      remove_inactive_deployments(active_deployment_names)
+    end
+
+    def sync_deployment_state(deployment_name, instances_data)
+      sync_instances(deployment_name, instances_data)
+      sync_agents(deployment_name, get_instances_for_deployment(deployment_name))
+    end
+
+    def sync_instances(deployment_name, instances_data)
+       deployment = @deployment_name_to_deployments[deployment_name]
+       active_instance_ids = sync_active_instances(deployment, instances_data)
+       remove_inactive_instances(active_instance_ids, deployment)
+    end
+
+    def sync_agents(deployment_name, instances)
+      deployment = @deployment_name_to_deployments[deployment_name]
+      active_agent_ids = sync_active_agents(deployment, instances)
+      remove_inactive_agents(active_agent_ids, deployment)
+      update_rogue_agents(active_agent_ids)
+    end
+
+    def remove_deployment(name)
+      deployment = @deployment_name_to_deployments[name]
+      deployment.agent_ids.each { |agent_id| @rogue_agents.delete(agent_id) }
+      @deployment_name_to_deployments.delete(name)
+    end
+
+    def remove_agent(agent_id)
+      @logger.info("Removing agent #{agent_id} from all deployments...")
+      @rogue_agents.delete(agent_id)
+      @deployment_name_to_deployments.values.each { |deployment| deployment.remove_agent(agent_id) }
+    end
+
+    def get_instances_for_deployment(deployment_name)
+      @deployment_name_to_deployments[deployment_name].instances
+    end
+
+    def analyze_agents
+      @logger.info "Analyzing agents..."
```

Comment:
We'll clean this up when we merge but we prefer to use parenthesis whenever possible for clarity and consistency. Some exceptions might be in RSpec's idiomatic syntax (`expect(subject).to eq(something)`) but in this case we'd prefer explicit parents.